File: backend/apps/tasks/schemas.py
from datetime import date, datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from ninja import Schema
from pydantic import Field


class UserSummarySchema(Schema):
    id: int
    email: str
    first_name: str
    last_name: str


class TaskCreateSchema(Schema):
    project_id: UUID | None = None
    milestone_id: UUID | None = None
    parent_id: UUID | None = None
    assignee_id: int | None = None

    title: str = Field(min_length=1, max_length=250)
    description: str = ""

    status: str = "todo"
    priority: str = "normal"

    start_date: date | None = None
    due_date: date | None = None

    estimated_hours: Decimal = Field(
        default=Decimal("0.00"),
        ge=0,
    )
    actual_hours: Decimal = Field(
        default=Decimal("0.00"),
        ge=0,
    )

    progress: int = Field(default=0, ge=0, le=100)
    sort_order: int = Field(default=0, ge=0)

    labels: list[str] = Field(default_factory=list)

    is_recurring: bool = False
    recurrence_rule: str = Field(
        default="",
        max_length=255,
    )


class TaskUpdateSchema(TaskCreateSchema):
    pass


class TaskAssigneeCreateSchema(Schema):
    user_id: int


class TaskAssigneeSchema(Schema):
    id: UUID
    user: UserSummarySchema
    created_at: datetime


class TaskWatcherCreateSchema(Schema):
    user_id: int


class TaskWatcherSchema(Schema):
    id: UUID
    user: UserSummarySchema
    created_at: datetime


class TaskChecklistCreateSchema(Schema):
    title: str = Field(min_length=1, max_length=250)
    sort_order: int = Field(default=0, ge=0)


class TaskChecklistToggleSchema(Schema):
    is_completed: bool


class TaskChecklistSchema(Schema):
    id: UUID
    title: str
    is_completed: bool
    completed_at: datetime | None
    completed_by: UserSummarySchema | None
    sort_order: int
    created_at: datetime


class TaskCommentCreateSchema(Schema):
    content: str = Field(min_length=1)
    is_internal: bool = True


class TaskCommentSchema(Schema):
    id: UUID
    content: str
    is_internal: bool
    author: UserSummarySchema | None
    created_at: datetime


class TaskDependencyCreateSchema(Schema):
    related_task_id: UUID
    dependency_type: str = "related"


class TaskDependencySchema(Schema):
    id: UUID
    related_task_id: UUID
    related_task_title: str
    dependency_type: str
    created_at: datetime


class TaskTimeLogCreateSchema(Schema):
    user_id: int | None = None
    work_date: date
    hours: Decimal = Field(gt=0)
    description: str = ""
    is_billable: bool = True


class TaskTimeLogSchema(Schema):
    id: UUID
    user: UserSummarySchema
    work_date: date
    hours: Decimal
    description: str
    is_billable: bool
    created_at: datetime


class TaskEventSchema(Schema):
    id: UUID
    event_type: str
    description: str
    metadata: dict[str, Any]
    created_at: datetime


class TaskSchema(Schema):
    id: UUID

    project_id: UUID | None
    project_title: str | None

    milestone_id: UUID | None
    milestone_title: str | None

    parent_id: UUID | None

    title: str
    description: str
    status: str
    priority: str

    assignee: UserSummarySchema | None

    start_date: date | None
    due_date: date | None
    completed_at: datetime | None

    estimated_hours: Decimal
    actual_hours: Decimal

    progress: int
    sort_order: int
    labels: list[str]

    is_recurring: bool
    recurrence_rule: str

    additional_assignees: list[TaskAssigneeSchema]
    watchers: list[TaskWatcherSchema]
    checklist_items: list[TaskChecklistSchema]
    comments: list[TaskCommentSchema]
    dependencies: list[TaskDependencySchema]
    time_logs: list[TaskTimeLogSchema]
    events: list[TaskEventSchema]

    created_at: datetime
    updated_at: datetime

File: apps/dashboard/src/features/tasks/schemas.ts
import {
  z,
} from "zod";

import {
  taskPriorities,
  taskStatuses,
} from "./types";

const userSummarySchema = z.object({
  id: z.number().int(),
  email: z.string(),
  first_name: z.string(),
  last_name: z.string(),
});

const taskAssigneeSchema = z.object({
  id: z.string().uuid(),
  user: userSummarySchema,
  created_at: z.string(),
});

const taskWatcherSchema = z.object({
  id: z.string().uuid(),
  user: userSummarySchema,
  created_at: z.string(),
});

const taskChecklistSchema = z.object({
  id: z.string().uuid(),
  title: z.string(),
  is_completed: z.boolean(),
  completed_at: z.string().nullable(),
  completed_by: userSummarySchema.nullable(),
  sort_order: z.number().int().nonnegative(),
  created_at: z.string(),
});

const taskCommentSchema = z.object({
  id: z.string().uuid(),
  content: z.string(),
  is_internal: z.boolean(),
  author: userSummarySchema.nullable(),
  created_at: z.string(),
});

const taskDependencySchema = z.object({
  id: z.string().uuid(),
  related_task_id: z.string().uuid(),
  related_task_title: z.string(),
  dependency_type: z.string(),
  created_at: z.string(),
});

const taskTimeLogSchema = z.object({
  id: z.string().uuid(),
  user: userSummarySchema,
  work_date: z.string(),
  hours: z.string(),
  description: z.string(),
  is_billable: z.boolean(),
  created_at: z.string(),
});

const taskEventSchema = z.object({
  id: z.string().uuid(),
  event_type: z.string(),
  description: z.string(),
  metadata: z.record(
    z.string(),
    z.unknown(),
  ),
  created_at: z.string(),
});

export const taskSchema = z.object({
  id: z.string().uuid(),
  project_id: z.string().uuid().nullable(),
  project_title: z.string().nullable(),
  milestone_id: z.string().uuid().nullable(),
  milestone_title: z.string().nullable(),
  parent_id: z.string().uuid().nullable(),
  title: z.string(),
  description: z.string(),
  status: z.enum(taskStatuses),
  priority: z.enum(taskPriorities),
  assignee: userSummarySchema.nullable(),
  start_date: z.string().nullable(),
  due_date: z.string().nullable(),
  completed_at: z.string().nullable(),
  estimated_hours: z.string(),
  actual_hours: z.string(),
  progress: z.number().int().min(0).max(100),
  sort_order: z.number().int().nonnegative(),
  labels: z.array(z.string()),
  is_recurring: z.boolean(),
  recurrence_rule: z.string(),
  additional_assignees: z.array(
    taskAssigneeSchema,
  ),
  watchers: z.array(taskWatcherSchema),
  checklist_items: z.array(
    taskChecklistSchema,
  ),
  comments: z.array(taskCommentSchema),
  dependencies: z.array(
    taskDependencySchema,
  ),
  time_logs: z.array(taskTimeLogSchema),
  events: z.array(taskEventSchema),
  created_at: z.string(),
  updated_at: z.string(),
});

export const paginatedTasksSchema = z.object({
  items: z.array(taskSchema),
  pagination: z.object({
    page: z.number().int().positive(),
    page_size: z.number().int().positive(),
    total_items: z.number().int().nonnegative(),
    total_pages: z.number().int().nonnegative(),
  }),
});
